[Incogdino] iP

Incogdino opened the pull request:
# David
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

David is a task manager application that helps you to remember your tasks. It is, 
- [x] text-based
- [x] easy to learn
- [x] ~~FAST~~ SUPER FAST to use

**The best part of it all, IT IS FREE**

You can use it to,
- Create a todo, deadline and event task
- Add or delete tasks
-  Save tasks locally 

To download David, 
1. Download the jar file from [here](https://github.com/Incogdino/ip). 
2. Run `gradlew shadowJar`
3. Double click on the jar file
4. Enjoy! 😄 

---
If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public static void main(String[] args) {
        new David("./src/main/java/David/Data/database.txt").activateChatBot();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Incogdino committed:
Added Level 0: Rename, Greet, Exit

Incogdino committed:
Added Echo (Duke Level - 1)

Incogdino committed:
Fixed small typo bug at constructor

Incogdino committed:
Added Add,List - level 3

Incogdino committed:
Added Level 3, Mark as Done

Incogdino committed:
Add Level 4. ToDos, Events, Deadlines

Incogdino committed:
Fix syntax error

Incogdino committed:
Update gitignore to ignore compiled class files

Incogdino committed:
Add automated Text UI testing, A-TextUiTesting

Incogdino committed:
Add error handling

Incogdino committed:
Add delete function to allow users to delete a task from the list

Incogdino committed:
Add JavaDoc comments

Lack of proper JavaDoc comments for some methods.

Incogdino committed:
Add local cache feature

A local cache save allows users to to save the list of items added
in their local machine. Retrieval of the list upon relaunching the
application returns the saved list.

Incogdino committed:
Add cache exception

A more appropriate error message is displayed when the specified
path for storing the cache file does not exist or is invalid.

Users who get the error message are prompted to report the bug to
the maintainers.

Incogdino committed:
Merge branch 'branch-Level-7'

Incogdino committed:
Add DateParser class

The DateParser class parses a given string into a LocalDate class.

Incogdino committed:
Add DateParser class

The DateParser class parses a given string into a LocalDate class.

Incogdino committed:
Change date and time to LocalDateTime rather than plain strings

Date and time are handled using LocalDateTime class to harness
Date and time methods supplied by the LocalDateTime library.

Incogdino committed:
Add invalid date time exception

Added proper error handling with appropriate error messages when
the date and time supplied is invalid.

Incogdino committed:
Merge branch 'branch-Level-8'

Incogdino committed:
Add more OOP style

Abstracted details from the main class and separated their concerns
into different packages and files.

Incogdino committed:
Refactor code to follow coding standard.

Incogdino committed:
Revert "Refactor code to follow coding standard."

This reverts commit 2b96c75d2d1518179b45b97261e016b7af7bbbc2.

Incogdino committed:
Refactor code to follow coding standard

Incogdino committed:
Merge branch 'A-MoreOOP'

Incogdino committed:
Refactor files into packages

All files including main file is added into a package. Packages
helps in better file management of the codebase.

Incogdino committed:
Merge branch 'branch-A-Packages'

Incogdino committed:
Merge branch 'add-gradle-support'

Incogdino committed:
Add gradle support

Gradle can be used to automate build tasks of the project.

Incogdino committed:
Refactor DateTask to handle errors instead

Separated concerns to allow DateParser class to handle the error
instead.

Incogdino committed:
Add test for DateParser and StringParser class

DateParser and StringParser class are not tested.

Test them with edge cases and special cases

Incogdino committed:
Merge branch 'branch-A-JUnit'

Incogdino committed:
Update Storage to create new cache file with relative path

Using absolute path created caching errors when running the JAR
file.

Creating a relative path allowed for the file to be created
wherever the JAR file was ran.

Incogdino committed:
Add JavaDoc comments

JavaDoc comments helps to explain what the methods do and the
exceptions it can throw. It provides greater clarity for the
methods.

Incogdino committed:
Merge branch 'branch-A-JavaDoc'

Incogdino committed:
Refactor to follow Coding standard

Better consistency in the code written for different parts.

Incogdino committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/david/Task/DeadlineTask.java
#	src/main/java/david/Task/EventTask.java

Incogdino committed:
Add find action to search for an event

Users can search for an event by appending "find" with any event
name.

Incogdino committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/david/Data/Storage.java
#	src/main/java/david/David.java
#	src/main/java/david/UI/Ui.java

Incogdino committed:
Refactor Storage class to follow coding standard

Incogdino committed:
Refactor code to follow java coding conventions

Following predefined java coding conventions makes code more
uniform and standardize and thus easier to read and understand.

Incogdino committed:
Refactor test classes to fit coding convention

Following predefined java coding conventions makes code more
uniform and standardize and thus easier to read and understand.

Incogdino committed:
Merge branch 'branch-A-CheckStyle'

Incogdino committed:
Add JavaFx GUI for better user experience

Current CLI user interface is not very pleasing. A GUI is added using
JavaFx to provide a better user experience.

Incogdino committed:
Merge branch 'branch-Level-10'

leroychiu20 commented on src/main/java/david/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Abstract task class
+ */
+public abstract class Task {
```
I like how you made the Task class an abstract class since it wouldn't be instantiated.

leroychiu20 commented on src/main/java/david/ui/Ui.java:
```diff
@@ -0,0 +1,142 @@
+ */
+public class Ui {
+    private static final String INTRO =
+            "____________________________________________________________\n"
```
Perhaps you could consider initialising a constant 'LINE' since this line is used a couple of times in your code

leroychiu20 commented on src/main/java/david/data/Storage.java:
```diff
@@ -0,0 +1,143 @@
+package david.data;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+import david.exceptions.DavidCacheException;
+import david.exceptions.DavidInvalidDateTimeException;
+import david.parser.DateParser;
+import david.task.DeadlineTask;
+import david.task.EventTask;
+import david.task.Task;
+import david.task.TaskList;
+import david.task.TodoTask;
```
I like how there is a consistent ordering of import statements which makes it easier to browse the list and determine the dependencies when there are many imports, and also how you didn't use the wildcard for importing.

leroychiu20 commented on src/main/java/david/task/TaskList.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    public String findEvent(String s) {
+        String events = " ____________________________________________________________\n"
+                + "Here are the matching tasks  in your list:\n";
```
I think there is an extra space between 'tasks' and 'in'?

Incogdino replied:
That is a good catch thanks for this!

leroychiu20 commented on src/main/java/david/ui/Ui.java:
```diff
@@ -0,0 +1,142 @@
+     * Displays successful unmarking of a task
+     * @param t Task to unmark as done
+     */
+    public String displayMarkAsUnDoneMessage(Task t) {
```
should the method name be displayMarkAsUndoneMessage(Task t) instead?

leroychiu20 reviewed (commented):
LGTM. Code was easy to read, just added a few nits.

samriddh2145 commented on src/main/java/david/David.java:
```diff
@@ -0,0 +1,239 @@
+ * Main chatbot class
+ */
+public class David {
+    private TaskList tasks;
+    private Ui ui;
+    private String inputString = "";
+    private Storage cache;
```
Good work with variable names.

samriddh2145 commented on src/main/java/david/MainWindow.java:
```diff
@@ -0,0 +1,86 @@
+        InputStream inputStream = this.getClass().getResourceAsStream(path);
+        try {
+            if (inputStream == null) {
+                throw new FileNotFoundException();
```
Should handle this exception

samriddh2145 commented on src/main/java/david/parser/DateParser.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class DateParser extends Parser {
+
+    /**
+     * Takes an input date time format in the form of " yyyy-MM-dd HHmm" and parses
+     * it into a LocalDateTime instance.
+     * @param date String date in " yyyy-MM-dd HHmm"
+     * @return LocalDateTime instance
+     * @throws DavidInvalidDateTimeException
+     */
```
Good job on enhancing readability through detailed comments.

samriddh2145 commented on src/main/java/david/task/TaskList.java:
```diff
@@ -0,0 +1,104 @@
+     * @param s String specified by the user
+     * @return  String containing all events matching the string specified by the user
+     */
+    public String findEvent(String s) {
+        String events = " ____________________________________________________________\n"
+                + "Here are the matching tasks  in your list:\n";
+        for (int i = 0; i < tasks.size(); i++) {
+            Task t = tasks.get(i);
+            String[] eventDetails = t.getTask().split(" ");
+            List<String> eventArr = new ArrayList<>(Arrays.asList(eventDetails));
+            if (eventArr.contains(s)) {
+                events += t + "\n";
+            }
+        }
+
+        events += " ____________________________________________________________\n";
+        return events;
+    }
```
Maybe return empty list instead of string when no tasks are found

samriddh2145 reviewed (commented):
Overall, good job with the implementation of tasks and adherence to naming conventions.

Incogdino committed:
Refactor code to fit java coding standard

Following predefined java coding conventions makes code more
uniform and standardize and thus easier to read and understand.

Incogdino committed:
Add assertions to David and Task class

Assertions helps ensure invariants assumed hold at certain parts of the
code.

Incogdino committed:
Refactor code to improve code quality

Better code quality can improve readability of the code. By making the
code more logical and easier to understand.

The code adheres to a logical structure. Magic numbers have been
drastically reduced/ removed. Complicated functions have been extracted
into separate functions adhering to SLAP principle. KISS principle have
also been applied to keep code easy to understand.

Incogdino committed:
Add assertions to Storage class

Assertions helps ensure invariants assumed hold at certain parts of the
code. This reduces the time needed to debug since assertions return
true at certain parts of the code.

Incogdino committed:
Fix test error for DateParser

DateParser returned took in wrong input String format. Proper date time
input format was changed to fix the error.

Incogdino committed:
Merge pull request #2 from Incogdino/branch-A-Assertions

Add assertions to David and Task class

Incogdino committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/david/David.java

Incogdino committed:
Merge pull request #1 from Incogdino/branch-A-CodeQuality

Refactor code to improve code quality

Incogdino committed:
Add sorting to allow tasks to be sorted in ascending or descending order

Sorting allows users to be able to arrange their tasks in the order they
desire. This increases the user experience and handles all the ordering
of tasks to the application.

Incogdino committed:
Merge branch 'BCD-Extension'

Incogdino committed:
Add github actions

Github actions helps to set up Continuous Integration.

Incogdino committed:
Remove manual addition of gradle.yml file

Incogdino committed:
Create gradle.yml

Incogdino committed:
Add README details

Provided more details on what David can do. A more user friendly
README is provided to allow users to have a better user experience
when trying out the application.

Incogdino committed:
Merge branch 'master' of https://github.com/Incogdino/ip

Incogdino committed:
Fix checkstyle errors

Incogdino committed:
Update user guide

The previous user guide lacked details in the final screenshot of
the application.

Incogdino committed:
Update user guide

The previous user guide lacked details in the final screenshot of
the application.

Incogdino committed:
Add more personality to David chatbot

Interaction with David felt monotonous and lacking personality.
A chatbot with better personality can enhance user interaction and
retention. It also provides a more personal user experience.

Incogdino committed:
Merge branch 'branch-A-Personality'

Incogdino committed:
Refactor code to fit java coding standard

Incogdino committed:
Fix jar file not working

Jar file had troubles launching due to a typo in the url link to the
image.

Incogdino committed:
Fix file not caching locally issue

File was not caching locally when ran with the jar file. Local caching
helps to save the user's tasks so that they can retrieve them again
on relaunching the application.

Incogdino committed:
Refactor Javadoc comments to meet java coding standard

Incogdino committed:
Update ReadME for github repo

Incogdino committed:
Update README to show images

Incogdino committed:
Update README to show images

Incogdino committed:
Update README to show images

Incogdino committed:
Add more specific error handling to David class

There was no check to ensure that the inputted dates are valid, date
inputted should be after the current time. For event tasks, "/to" date
should be after the "/from" date.

Incogdino committed:
Refactor code to fit java coding standard

Better code quality can improve readability of the code. By making the
code more logical and easier to understand.

The code adheres to a logical structure. Magic numbers have been
drastically reduced/ removed. Complicated functions have been extracted
into separate functions adhering to SLAP principle. KISS principle have
also been applied to keep code easy to understand.

Incogdino committed:
Add tests to stress test DateParser methods

Incogdino committed:
Add help function for better user experience

The application lacked a help function to inform users of what commands
were available. A help function can provide quick access to a list of
commands that are accepted by the application, improving user's
experience.

Incogdino committed:
Add trim function to find command for better search output

Trimming leading and trailing whitespace can enhance the search output
in the case where the user adds whitespaces unintentionally causing the
find command to return nothing.

Incogdino committed:
Update README files to display help command feature

Incogdino committed:
Update findEvent method to better match user input string

The findEvent method is able to better match user input string by
splitting the string into its individual keyword and return the event
when a keyword is matched

Incogdino committed:
Update README with installation instructions

Installation instructions were missing in the README directive.

Including the installation instructions can help to increase user
experience especially for first time users.

Incogdino committed:
Add tests to DateParser class

DateParser method lacked sufficient testing for the correct output on
edge cases. Doing so lowers the chance of bugs and introduces unit
testing on the methods.

More test is done using edge cases of the DateParser class.

Incogdino committed:
Refactor validateDateTime method of the DateParser class

The method failed test cases where the input date time is equal to the
current date time. As such, a false positive result was produced.

The method now checks if the input time is equal to the current time and
returns the right output for it.

Incogdino committed:
Add tests to DateParser class

DateParser method lacked sufficient testing for the correct output on
edge cases. Doing so lowers the chance of bugs and introduces unit
testing on the methods.

More test is done using edge cases of the DateParser class, specifically
the case when the input date time are the same.

Incogdino committed:
Update find command to handle duplicates being output

Find command returns duplicates of the same task.

Duplicates can confuse the user and affect user experience with the
application.

Incogdino committed:
Fix README not displaying on deployed page error

README on the deployed github page not getting loaded correctly,
resulting in raw text in GFMD to be presented instead of the
formatted text.

Incogdino committed:
Fix README not displaying on deployed page error

README on the deployed github page not getting loaded correctly,
resulting in raw text in GFMD to be presented instead of the
formatted text.

Incogdino committed:
Revert "Fix README not displaying on deployed page error"

This reverts commit 0d3d240e2e40d200e99279a39da266c4d095bfbb.

Incogdino committed:
Fix README not displaying on deployed page error

README on the deployed github page not getting loaded correctly,
resulting in raw text in GFMD to be presented instead of the
formatted text.

Incogdino committed:
Fix README not displaying on deployed page error

README on the deployed github page not getting loaded correctly,
resulting in raw text in GFMD to be presented instead of the
formatted text.

Incogdino committed:
Fix README not displaying on deployed page error

Github pages uses kramdown to render the pages which leads to formatting
errors.

Incogdino committed:
Fix README not displaying on deployed page error

Github pages uses kramdown to render the pages which leads to formatting
errors.

Incogdino committed:
Fix README not displaying on deployed page error

Github pages uses kramdown to render the pages which leads to formatting
errors.

Incogdino committed:
Fix README not displaying on deployed page error

Github pages uses kramdown to render the pages which leads to formatting
errors.

Incogdino committed:
Change theme and markdown render

Test to see if a different markdown renderer will render alerts in GFMD
correctly on github pages.

Incogdino committed:
Revert markdown renderer to use GFM

Previous markdown renderer seem to load GFMD better instead.

Incogdino committed:
Revert markdown renderer to get false output

Required an example for posting an issue in the forum.

Incogdino committed:
Revert "Revert markdown renderer to get false output"

This reverts commit b3d100b6e0cbbf29342acea8579c95797f312c76.

Incogdino committed:
Revert markdown renderer to get false output

Required an example for posting an issue in the forum.

Incogdino committed:
Fix README not displaying on deployed page error

Github pages uses kramdown to render the pages which leads to formatting
errors.

Incogdino committed:
Refactor README to use pipe characters

HTML tables causes rendering issues for some markdown processors,
causing the table to be rendered wrongly.

Incogdino committed:
Remove duplicate dependencies in build.gradle

Duplicated junit dependencies caused issues running jar file in
macOS systems. Removing the extra dependencies caused the errors
to go away.

Incogdino committed:
Fix typo error for invalid time error message

Previous error message used the backslash instead of the forward
slash which may cause users to misinterpret and assume backward
slash should be used instead of a forward slash.

Incogdino committed:
Add ability to add tasks past current date time

Users should have the ability to add tasks even if the event have
passed for tracking purposes.

Incogdino committed:
Refactor java test methods to follow naming convention

Better naming convention for test methods can help to greatly
improve code readability and understanding. This makes it more
maintainable in the long run.

Incogdino committed:
Add more error handling for invalid index range inputs

Previously inputting negative index ranges for delete, mark and
unmark commands led to unexpected errors occurring such as app not
responding or outright crashing.

Proper error handling should be ensured to make sure the
application works as expected and the user receives proper feedback
on their input.
